// mailer/lib/mailer.js
var nodemailer = require('nodemailer'),
    ejs  = require('./ejs'),
    fs   = require('fs'),
    path = require('path'),

    templates = {};

exports.sendEmail = sendEmail;
exports.init = initMailer;

function initMailer () {
    try {
        var settings = require('yaml').eval(require('fs').readFileSync(app.root + '/config/mailer.yml').toString('utf8'))[app.settings.env];
    } catch (e) {
        console.log('Could not init mailer extension, env-specific settings not found in config/mailer.yml');
        console.log('Error:', e.message);
        return;
    }
    if (!settings) {
        return;
    }
    exports.settings = settings;

    switch (settings.mailer) {
    case 'sendmail':
        nodemailer.sendmail = true;
        break;
    case 'smtp':
        nodemailer.SMTP = {
            host: settings.host || "localhost",
            port: settings.port || 25,
            use_authentication: settings.use_authentication || false,
            user: settings.user || '',
            pass: settings.pass || ''
        };
        break;
    }

    // read app/views/emails dir
    var emailsDir = app.root + '/app/views/emails';
    if (path.existsSync(emailsDir)) {
        fs.readdirSync(emailsDir).forEach(function (file) {
            templates[file] = fs.readFileSync(emailsDir + '/' + file).toString('utf8');
        });
    }
}

function sendEmail (template, data, options) {
    options = options || {};
    data = data || {};
    data.url = exports.settings.url;

    var html = ejs.tmpl(getTemplate(template, 'html', options.locale), data),
        body = ejs.tmpl(getTemplate(template, 'test', options.locale), data),
        self = this,
        email = options.email;

    nodemailer.send_mail({
        sender:  options.from || exports.settings.from,
        to:      options.email,
        subject: options.subject,
        body:    body,
        html:    html
    }, function (err, success) {
        if (success) {
            console.log('=== Email sent');
            console.log(options.subject);
            console.log(body);
        } else {
            console.log(err);
        }
    });
}

function getTemplate (template, format, locale) {
    var fullName = [template, format, locale].join('.'), // register.en.html
        formatName = [template, format].join('.');       // register.html

    return templates[fullName] || templates[formatName] || ' missing template ' + formatName + (locale ? ' or ' + fullName : '');
}
